### ISSUE TITLE
rule to add wood care treatement as suggested product

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#920]
_Imported from trac ticket 920.
Created by **rneville**
Opened in trac: May 27, 2011, 10:21:18 AM
Last modified in trac: Dec 12, 2012, 3:15:05 PM_
***
_by rneville:_
    12.   We need to show the wood care warranty sku 394000004 in related items when there are wood products on an order.

Populate ITM2RSKU appropriately or provide us with minor codes, etc. to make a rule from. 

12 a.       Any time an item is sold where ITM_CD  IN ITM$ITM_SRT WHERE ITM_SRT_CD = ‘WUV’ the user should be prompted to add the wood kit ‘394000005’

 

12 b.      Any time an item is sold where ITM_CD  IN ITM$ITM_SRT WHERE ITM_SRT_CD = ‘LUV’ the user should be prompted to add the lacquer kit ‘391000002’

**Comment by kbower** [comment]
_**kbower** trac comment on Jun 4, 2011, 10:06:50 AM:_

Should be another OR so there is not an extra database round trip.

**Comment by kbower** [comment]
_**rneville** trac comment on Jun 14, 2012, 2:33:09 PM:_

Calvin, see one of us for explanation before starting on this...

**Comment by kbower** [comment]
_**criley** trac comment on Oct 17, 2012, 11:18:48 AM:_

(Changeset [[3273]](../commit/784cdfc4d6f889795540e8327881a2d1d0b2e6ee))
*   Fixes #920 (rule to add wood care treatment as suggested product)

**Comment by kbower** [comment]
_**criley** trac status change on Oct 17, 2012, 11:18:48 AM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 18, 2012, 7:23:17 AM:_

Calvin, you are certainly on the right track here.  But the product.recordgroups collection will be a collection of `ProductRecordGroup` objects, not strings (test this by adding one using '`paster shell`').

So, in other words, this if will never evaluate to True:
```python
if 'WUV' in product.recordgroups:
    ...
```
Take a look at the '`obj_with_attr`' function in `utils.py`.
You could use something like this:
```python
if obj_with_attr(product.recordgroups, 'recordgroupid', u'WUV'):
    ...
```
However, for Kane's especially we need to be mindful of the number of database trips, especially for common http requests, like this one (this might be the single most common http request).  When sqlalchemy hits '`product.recordgroups`', it needs to make another database trip in order to look up the recordgroups.

So I'd recommend we use an "EXISTS" clause here, like is done with Kan3YearWarranty right in that same rule.

The u'WUV' portion would like this, but you'll need to set up a little data and test/fix it before committing:
```python
subq=DBSession.query(ProductProductRecordGroup.recordgroupid)\\
    .filter(ProductProductRecordGroup.recordgroupid==u'WUV')\\
    .filter(ProductProductRecordGroup.productid==product.productid)\\
    .statement
or_list.append(and_(Product.productid==u'394000005', exists(subq)))
```
That will skip the final table and use only the secondary join table so we avoid a join and it would add SQL something like this:
```
OR products.productid = '394000005' AND (EXISTS (SELECT productproductrecordgroups.recordgroupid
FROM productproductrecordgroups
WHERE productproductrecordgroups.recordgroupid = 'WUV' AND productproductrecordgroups.productid = 'MYPRODUCT'))
```
Whichever result is faster is the one we should use.

**Comment by kbower** [comment]
_**kbower** trac status change on Oct 18, 2012, 7:23:17 AM:_ **reopened**

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 18, 2012, 7:24:25 AM:_

Also, as long as you're in that rule, can you please make the hard coded string '393000004' in that rule a unicode: u'393000004'

Thanks, man!

**Comment by kbower** [comment]
_**kbower** trac comment on Dec 4, 2012, 11:45:03 AM:_

(Changeset [[3458]](../commit/971b7d9706fc0e230854c0167de2b7ce46f7a70d))
* revert [r3273](../commit/784cdfc4d6f889795540e8327881a2d1d0b2e6ee) (see #920 comments)

**Comment by kbower** [comment]
_**criley** trac comment on Dec 12, 2012, 3:15:05 PM:_

(Changeset [[3492]](../commit/468ebdf4849b395a3748c944147a033a5ba34329))
* Fixes #920 (rule to add wood care treatment)

**Comment by kbower** [comment]
_**criley** trac status change on Dec 12, 2012, 3:15:05 PM:_ **closed**
